torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -685,51 +747,105 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(
         // aliasing to inputs, no need to allocate real output
-        outputs.push_back(
-            inferAndAlloc(output, {}, expr_eval, {}, options_, false));
+        // but we still need to push an entry here.
+        ret.push(int64_t(0));
       } else {
-        // Allocate a real output
-        outputs.push_back(
-            inferAndAllocOutput(output, expr_eval, options_, false));
+        // TODO: we are using meta here, which is bad since it doesn't account for devices. Switch to fake tensor instead
+        ret.push(inferAndAllocOutput(output, expr_eval, meta_options, false));
       }
     }
   }
-  return outputs;
+  return ret;
 }
 
-void FusionExecutor::setUsedTVs() {
-  auto used_vals = fusion_->usedMathVals();
-  auto used_tvs = ir_utils::filterByType<TensorView>(used_vals);
-  used_tvs_.clear();
+KernelArgumentHolder FusionExecutor::inferOutputSizes(
+    const KernelArgumentHolder& args,
+    const LaunchParams& launch_constraints) {
+  FUSER_PERF_SCOPE("FusionExecutor::RunFusion");
 
-  for (auto tv : used_tvs)
-    used_tvs_.push_back(tv);
+  ExecutorEntry* executor_entry = nullptr;
+  c10::optional<size_t> opt_code = args.getCacheId();
+  if (opt_code.has_value()) {
+    executor_entry = &executor_entry_lookup_[*opt_code];
+  }
+
+  executor_utils::initializeCudaContext();
+  TORCH_INTERNAL_ASSERT(lowered_);
+
+  TORCH_INTERNAL_ASSERT(
+      !executor_entry || !executor_entry->init,
+      "compile kernel shouldn't hit a pre-existing cache");
+  FUSER_PERF_SCOPE("ExecutorRunFusion::ValidateAndInitialize");
+  // TODO: validate kernel inputs currently won't be happy, since our fusion args are mapped with `meta` tensor instead of `cuda` tensor, check if this would be resolved with FakeTensor
+  // executor_utils::validateKernelInputs(fusion_, args, options_.device);
+
+  if (!evaluator_precomputed_integers_) {
+    evaluator_precomputed_integers_ =
+        std::make_unique<KernelPrecomputedIntegers>(lowered_->kernel());
+  }
+
+  kir::ExpressionEvaluator expr_eval;
+  evaluator_precomputed_integers_->bindKernelInputs(lowered_->kernel(), args);
+  expr_eval.precomputedIntegers() = evaluator_precomputed_integers_.get();
+
+  // I think this binds something to expr_eval, so even though we are not using
+  // launch_params_, we still need this in order to infer output shapes.
+  launch_params_ =
+      computeLaunchParams(launch_constraints, expr_eval, warp_size_);
+
+  executor_utils::validateVectorizedTensors(
+      lowered_.get()->kernel(), args, {}, compileTimeDataCache(), expr_eval);
+
+  auto alias_indices_entry = executor_utils::caching::ExecutorCompileTimeEntry<
+      executor_utils::caching::InputAliasIndices>(
+      compileTimeDataCache(), [&]() {
+        return std::make_unique<std::vector<std::pair<int, int>>>(
+            fusion_->getInputAliasIndices());
+      });
+
+  auto& alias_indices = alias_indices_entry.get();
+
+  // ditch pre-allocated outputs if the number doesn't match.
```

Comment:
I don't follow this comment.

Reply:
errr... neither do I. Looking at the place where it's copied from I think it's obsolete, since we now error out when output doesn't match, instead of silently swapping it with fresh outputs.

Reply:
removed